By default, use the C compiler driver rather than calling the assembler directly

shindere opened the pull request:
This is a follow-up to @xavierleroy's comment https://github.com/ocaml/ocaml/pull/10063#pullrequestreview-564887724

It modifies the build system to go through the C compiler driver to invoke
the assembler by default, rather than invoking it directly.

It went through precheck, se [Build #539](https://ci.inria.fr/ocaml/precheck/build/539/).
This build has one failure but which is unrelated.

shindere committed:
configure: simplify assembler detection

By default, use the C compiler driver rather than calling the
assembler directly, except on some platforms where it is well known
and easy to call the assembler directly.

gasche reviewed (commented):
With this choice, in the vast majority of cases we have `default_as = default_aspp`, can use this property to simplify the definitions instead of duplicating the compiler invocation command?

Line 1102-1103, and also line 1108-1109, `as` and `aspp` are not defined in the same way. I assume these are mistakes/irregularities rather than an intentional difference? (Otherwise a comment could be nice.)

Finally: I would have expected the choice of `<c-compiler> -c` to vastly simplify the configure script, given that we already know what the correct `<c-compiler>` invocation should be, so there should be much less architecture-dependent conditionals to decide how to add the `-c` flag. Am I missing something?

dra27 commented on configure.ac:
```diff
@@ -1092,14 +1092,14 @@ AS_CASE(["$arch,$system"],
         [AS_IF([test x"$SYSTEM_AS" = "x"],
           [AC_MSG_ERROR([GNU as assembler is required.])],
-          [default_as="${toolpref}as --64"
+          [default_as="${toolpref}gcc -m64 -c"
```
In this case we tested for `as` above - don't we need to test for `gcc` as well? cf. difference between `default_as` and `default_aspp` here?

dra27 commented on configure.ac:
```diff
@@ -1092,14 +1092,14 @@ AS_CASE(["$arch,$system"],
+          [default_as="${toolpref}gcc -m64 -c"
           default_aspp="${toolpref}cc -m64 -c"])],
       [gcc-*],
         [AS_IF([test x"$SYSTEM_AS" = "x"],
```
I think this whole test (i.e. the use of `SYSTEM_AS` can disappear with this?

dra27 reviewed (commented):
It certainly looks like we should be able to eliminate `$default_aspp` - it might be possible to simplify that `AS_CASE` further by setting flags for the assembler at L395 (when `$arch`, `$system` and `$model` are set), adding `default_assembler_flags` or something (I'm not sure what the best name for "flags to the pass to the C compiler to pass the assembler" is!). Then only msvc need specific values for `default_as` to be specified.

shindere commented:
Many thanks for your feedback, @gasche ang @dra27.

So for those who are not familiar, the build system uses two commands to
assemble files: AS and ASPP.

The AS command is used to assemble files generated by ocamlopt, whereas
ASPP assembles handwritten files and does some preprocessing, too.

In my understanding, it's best to keep the distinction in the build
system between AS and ASPP, so that users have more control on what they
want to override.

Thus, the only thing this PR deals with so far is the default value for these
two variables. Until now this default value used to be different, and
this PR proposes to use the same default value, namely the C compiler
driver (so the with-preprocessing variant).

So the PR has currenlty two commits that do the folloing:

 1. The first commit makes sure the two variables have the same default value everywhere.

 2. The second commit removes the definition of `default_aspp` and uses `default_as` to initialize both `AS` and `ASPP`.

Before I continue, do we agree on these two commits?

I am also thinking aobut matching on the C compiler because if it's
clang then it seems to use basically the same flags on matter the other
variables.

What do you think?

gasche reviewed (approved):
Yes, with this new second commit the PR looks much nicer.

David commented on a test lines 1103-1105 that is now useless and can be removed, given that the two branches do the exact same thing.

With that minr issue fixed, I would think that the PR is ready to go. (I'm relying on the test results that you observed, rather than reasoning on the code change, to assume that it is probably a safe change.) A Changes entry would be nice.

shindere commented:
Thanks a lot for your feedback, @gasche.

Gabriel Scherer (2021/02/01 12:37 -0800):
> @gasche approved this pull request.

Thanks.

> Yes, with this new second commit the PR looks much nicer.

Good.

> David commented on a test lines 1103-1105 that is now useless and can
> be removed, given that the two branches do the exact same thing.

Ah yes that's correct indeed, thanks. This has now been removed.

> With that minr issue fixed, I would think that the PR is ready to go.
> (I'm relying on the test results that you observed, rather than
> reasoning on the code change, to assume that it is probably a safe
> change.) A Changes entry would be nice.

Added, too.

Looking at the code, I believe that the big case used to determine the
default value is not exhaustive. It was probably also the case before
this PR, butperhaps now is a good time to fix this and make sure
something meaningful is done in any case. If we want to do something, I
can see two approaches. Either we initialize `default_as`with a
reasonable default, such as `$CC -c`, or we initialize it to the empty
string before the case and give an error message if it's still empty
after the case.

Any opinion?

xavierleroy commented:
This looks like a nice simplification already.  However, what I had in mind was to use "whatever C compiler and flags we use for the rest of the system" plus the "-c" flag as default value for AS and ASPP.  Possibly with "-Wno-trigraph" added if the C compiler is clang.  Then, the only cases where this default is overriden would be
- win32 and win64, to call the Microsoft assembler instead;
- possibly arm64/arm/arm64/i386/riscv with gcc, where we know that `${toolpref}as` works, reducing compilation times a bit compared with going through the C compiler.

But I may have missed something here.

shindere commented:
Xavier Leroy (2021/02/06 09:32 -0800):
> This looks like a nice simplification already.  However, what I had in
> mind was to use "whatever C compiler and flags we use for the rest of
> the system" plus the "-c" flag as default value for AS and ASPP.
> Possibly with "-Wno-trigraph" added if the C compiler is clang.  Then,
> the only cases where this default is overriden would be
> - win32 and win64, to call the Microsoft assembler instead;

This is what is implemented in Experiment #2, which seems to work quite
well, see [precheck build #544](https://ci.inria.fr/ocaml/job/precheck/544/) (the failure on
MinGW64 is unrelated, due to a corruption of the workspace but I am
unable to fix it right now because at the moment I can't connect to the
slave in question).

Note that I based the detection on target rather than host. This
actually makes much more sense since it's for the target (and not for
the host) that we want to assemble files. I may even go one step further
and using `$target_alias` because that has proven useful in the context
of cross-compilation.

> - possibly arm64/arm/arm64/i386/riscv with gcc, where we know that
> `${toolpref}as` works, reducing compilation times a bit compared
> with going through the C compiler.

Not implemented yet. I am correct that your first "arm64" actually
intended to be "amd64"?

> But I may have missed something here.

I don't think so!

If you can confirm that you meant amd64 I may try to implement this. Am
I correct that this assumes Linux as an OS? Should we somehow make sure
the tool exist before adopting is as our default?

xavierleroy commented:
Yes, I meant "**amd64**/arm/arm64/i386/riscv with gcc" as the platforms that could deserve a special case.

mshinwell commented:
I am in favour of @xavierleroy 's proposed mitigation: I'm concerned that the performance impact of this change might be noticeable on large builds.  gcc, for example, has to execute the assembler separately since it isn't built in.  This probably won't double the amount of process creations per build (since there will be shell invocations etc. too), but it seems like it might increase them by half?  Indeed gcc also seems to search the path to find "as", so there will also be a lot of unnecessary calls to one of the "exec" functions.

On macOS we are already calling via clang, which appears (as far as I can tell; it's hard to trace these days with SIP enabled) from its `-v` output to be doing a separate execution of the assembler.

shindere commented:
Xavier Leroy (2021/02/11 09:54 -0800):
> Yes, I meant "**amd64**/arm/arm64/i386/riscv with gcc" as the
> platforms that could deserve a special case.

Working on this currently, I am realising that if we go this way, then
it is no longer possible to use the same value as default for as and
aspp, otherwise we are going to lose the preprocessing. I guess on these
platforms we should use ${toolpref}as as default for as and keep using
`$CC -c` as default for the value for ASPP.

shindere commented:
So, the PR has been updated with a new proposal which seems to work, see
[precheck build #546](https://ci.inria.fr/ocaml/job/precheck/546/)

It includes an enhancement on the way `toolpref` is computed: the target
takes precedence over the host if present. That proved to be useful in
the context of cross-compiling.

xavierleroy commented on INSTALL.adoc:
```diff
@@ -46,8 +46,7 @@ Examples:
     ./configure --build=x86_64-pc-linux-gnu --host=i386-linux \
-                CC='gcc -m32' AS='as --32' ASPP='gcc -m32 -c' \
-                PARTIALLD='ld -r -melf_i386'
+                CC='gcc -m32' PARTIALLD='ld -r -melf_i386'
```
This looks like a nice simplification.  However, I'm not sure it works.  Since the C compiler is gcc, and the target architecture i386, it looks like the configure script will take `default_as="${toolpref}as"`, missing the crucial `--32` flag that tells the assembler to produce 32-bit code.  I think you need to keep `AS='as --32'` here.  Getting rid of ASPP looks fine, though.

xavierleroy reviewed (commented):
This looks pretty nice.  Thanks!

One potential problem is below.

shindere commented:
Xavier Leroy (2021/02/17 10:30 -0800):
> @xavierleroy commented on this pull request.
>
> This looks pretty nice.  Thanks!

My pleasure! :)

> One potential problem is below.
>
>
> > @@ -46,8 +46,7 @@ Examples:
>  * On a Linux x86-64 host, to build a 32-bit version of OCaml:
>
>      ./configure --build=x86_64-pc-linux-gnu --host=i386-linux \
> -                CC='gcc -m32' AS='as --32' ASPP='gcc -m32 -c' \
> -                PARTIALLD='ld -r -melf_i386'
> +                CC='gcc -m32' PARTIALLD='ld -r -melf_i386'
>
> This looks like a nice simplification.  However, I'm not sure it
> works.  Since the C compiler is gcc,

Actually, configure enters cross-compilation mode and thus selects
i686-linux-gnu-gcc.

> and the target architecture i386,
> it looks like the configure script will take
> `default_as="${toolpref}as"`, missing the crucial `--32` flag that
> tells the assembler to produce 32-bit code.  I think you need to keep
> `AS='as --32'` here.  Getting rid of ASPP looks fine, though.

I just tried `./configure --build=x86_64-pc-linux-gnu
--host=i686-linux-gnu` and it worked just fine, as did the subsequent
`make`.

The values configure computes for AS and ASPP are:

```
ASM=i686-linux-gnu-as
ASPP=i686-linux-gnu-gcc -c
```

So, it seems to me the example in `INSTALL.adoc` could actually be
simplified even further. Or am I missing something?

xavierleroy commented:
> The values configure computes for AS and ASPP [in x86 32 bits mode] are:
> 
```
ASM=i686-linux-gnu-as
ASPP=i686-linux-gnu-gcc -c
```

OK, these are good values.  I wasn't expecting `${toolpref}` to be set in this case, but looks like it is.

This looks good overall.  I'm running another round of CI precheck to reassure me.

shindere commented:
Xavier Leroy (2021/02/22 08:58 -0800):
> > The values configure computes for AS and ASPP [in x86 32 bits mode] are:
> >
> ```
> ASM=i686-linux-gnu-as
> ASPP=i686-linux-gnu-gcc -c
> ```
>
> OK, these are good values.

So, I guess you are okay that I include an update to the INSTALL file,
right?

> I wasn't expecting `${toolpref}` to be set in this case, but looks
> like it is.

It's one of our variables, so we handleit as we want.

> This looks good overall.  I'm running another round of CI precheck to
> reassure me.

Good idea, thanks!

xavierleroy commented:
> So, I guess you are okay that I include an update to the INSTALL file, right?

The update that is already part of this PR looks OK.

I understand you want to simplify further.  Removing `CC='gcc -m32'` might work.  Not sure what happens if you remove `PARTIALLD='ld -r -melf_i386'`.  Did you test?

xavierleroy commented:
CI precheck is happy.

shindere commented:
Xavier Leroy (2021/02/22 09:29 -0800):
> > So, I guess you are okay that I include an update to the INSTALL file, right?
>
> The update that is already part of this PR looks OK.
>
> I understand you want to simplify further.  Removing `CC='gcc -m32'`
> might work.  Not sure what happens if you remove `PARTIALLD='ld -r
> -melf_i386'`.  Did you test?

Yes I did. It works perfectly well, as I thinkI wrote earlier when I
gave the computed values for AS and ASPP.

shindere commented:
Xavier Leroy (2021/02/22 09:52 -0800):
> CI precheck is happy.

Great, thanks. If you don't mind, I'll just update INSTALL, squash
everything together and merge tomorrow morning. Won't wait for
another green CI because I think this one used enough energy and will
gothrough CI again when merged anyway.

shindere commented:
As announced, I went ahead and:

 * Rebased on latest trunk
 * Further simplified the example in `INSTALL.adoc`
 * Squashed all the commits
 * Merged

Thanks a lot to all those who participated in this!
